_source/_posts/2020-07-02-spring-boot-using-java-modules.md

```diff
@@ -0,0 +1,465 @@
+* Enter `Spring Boot with Java Modules` for the **Name** (this value doesn't matter, so feel free to change it)
+* Change the Login redirect URI to be `http://localhost:8080/login/oauth2/code/okta`
+* Click **Done**
+
+### Configure the App with Okta Information
+
+Create a file `application/src/main/resources/application.properties` with the following content:
+
+```
+okta.oauth2.issuer=https://{yourOktaDomain}/oauth2/default
+okta.oauth2.clientId={clientId}
+okta.oauth2.clientSecret={clientSecret} 
+```
+
+You can find {clientId} and {clientSecret} in your application setup:  
+
+{% img blog/spring-boot-with-modules/okta-app.png alt:"Okta app setup" width:"542" %}{: .center-image }
+
+The {yourOktaDomain} you can find on Okta dashboard:  
+
+{% img blog/spring-boot-with-modules/org-url.png alt:"Okta Domain" width:"800" %}{: .center-image }
+
+Now if you restart the app and navigate to `http://localhost:8080/bird` you'll get a login page appearing.
+
+## Using Java Modules
+
+Now it is time to modularize the app. This is achieved by placing a file `module-info.java` on source root for each module. We'll be doing this for our two modules `application` and `persistence`. There are two ways to modularize a Java app: top-down and bottom-up. In this tutorial we'll be showing the bottom-up approach, that is modularizing the libraries before the app. This approach is preferable as we'll have `persistence` already modularized when writing the `application` `module-info.java`. If `application` was modularized first then `persistence` would be treated as an automatic module, and it would have to be referenced as the JAR file name.
+
+### Modularize `persistence` Library
+
+Create a module declaration file `persistence/src/main/java/module-info.java` with the following content:
+
+```java
+module com.okta.developer.modules.persistence {
+
+    requires java.annotation;
+    requires spring.beans;
+    requires spring.context;
+    requires spring.data.commons;
+    requires spring.data.mongodb;
+
+    exports com.okta.developer.animals.bird;
+}
+```
+
+Each `requires` keyword signalize that this module will be depending on some other module.
+Spring, on version 5, is not modularized yet so its JAR files don't have the `module-info.java`. 
+When you have a dependency on the `modulepath` (former classpath for non-modular applications) like this they will be available as `automatic modules`.
+
+An `automatic module` gets its name from the property `Automatic-Module-Name` inside `MANIFEST.MF` or from the `JAR` filename itself if that is absent.
+
+On Spring 5 the team was nice enough to put the `Automatic-Module-Name` for all the libraries. So those are the names used on our persistence app dependencies: `spring.beans`, `spring.context`, `spring.data.commons`, `spring.data.mongodb`.
+
+The `exports` keyword exports all classes in the package. When another module uses a `requires` clause referencing this one it will have access to those classes. 
```

Comment:
```suggestion
The `exports` keyword exports all classes in that package. When another module uses a `requires` clause referencing this one it will have access to those classes. 
```

Reply:
🤔